torch/csrc/jit/irparser.cpp

```diff
@@ -0,0 +1,446 @@
+#include <torch/csrc/jit/irparser.h>
+#include <torch/csrc/jit/ir.h>
+#include <torch/csrc/jit/script/lexer.h>
+#include <torch/csrc/jit/script/parse_string_literal.h>
+
+#include <string>
+#include <vector>
+
+namespace torch {
+namespace jit {
+namespace script {
+
+struct VarWithType;
+struct ParsedLiteral;
+
+class IRParser {
+  friend void parseIR(const std::string& str, torch::jit::Graph* graph);
+  IRParser(const std::string& str, torch::jit::Graph* graph)
+      : L(str), g(graph) {}
+
+  std::string parseVar();
+  VarWithType parseVarWithType();
+  ParsedLiteral parseScalarLiteral(Node* n);
+
+  void parse();
+  void parseGraphInputs();
+  void parseReturnOperator();
+
+  void parseBlocks(Node* parentNode);
+  void parseBlock(Node* parentNode);
+  void parseBlockInputs(Block* b);
+  void parseBlockOutputs(Block* b);
+
+  void parseOperatorsList(Block* b);
+  void parseOperator(Block* b);
+  void parseOperatorOutputs(std::vector<VarWithType>* outs);
+  std::string parseOperatorName();
+  void parseOperatorInputs(Node* n);
+  void parseAttrs(Node* n);
+  void parseAttr(Node* n);
+
+  void parseList(
+      int begin,
+      int sep,
+      int end,
+      const std::function<void()>& callback);
+
+  torch::jit::script::Lexer L;
+  torch::jit::Graph* g = nullptr;
+  std::unordered_map<std::string, Value*> vmap;
+};
+
+struct ParsedLiteral {
+  AttributeKind k = AttributeKind::t;
+
+  long i;
+  std::string s;
+  double f;
+  std::vector<long> is;
+  std::vector<std::string> ss;
+  std::vector<double> fs;
+};
+
+struct VarWithType {
+  VarWithType(){};
+  std::string name;
+  std::string type;
+};
+
+void parseIR(const std::string& str, torch::jit::Graph* graph) {
+  torch::jit::script::IRParser p(str, graph);
+  p.parse();
+}
+
+TypePtr parseType(const std::string& s) {
+  if (s == "Tensor" || s == "NoType") {
+    return TensorType::get();
+  }
+  if (s == "int") {
+    return IntType::get();
+  }
+  if (s == "float") {
+    return FloatType::get();
+  }
+  if (s == "string") {
+    return StringType::get();
+  }
+  // TODO: Support other types.
+  AT_ASSERTM(false, "Type not supported by parser:", s);
+}
+
+VarWithType IRParser::parseVarWithType() {
+  L.expect('%');
+  VarWithType r;
+  if (L.cur().kind == TK_IDENT) {
+    r.name = L.expect(TK_IDENT).text();
+  } else {
+    r.name = L.expect(TK_NUMBER).text();
+  }
+  r.type = "NoType";
```

Comment:
Probably don't need no type, can't it just be "Tensor" (based on usage in this file)